### ISSUE TITLE
setRemoteDescription and ontrack order of events: SRD before ontrack?

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-pc#1508]
TL;DR: Should `setRemoteDescription` be resolved before the resulting `ontrack` fires for remote tracks or after in the same task? @jan-ivar @taylor-b 

`setRemoteDescription` does, in parallel to returning a promise, a number of things. On success this includes:
```
1. Process remote tracks.
   1a. Queue a task to fire an event named track.
2. Resolve the promise.
```
As described, the promise would be resolved before `ontrack` fires. Invoking `then()` of a resolved promise means the resolver is not executed immediately, but [at the end of the current run of the JavaScript event loop](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#Timing), which is still before any tasks queued for the next JavaScript event loop. We get this order of events:
```
1. Synchronous code after our setRemoteDescription() call is executed as part of
   the current loop.
2. resolve is called (setRemoteDescription(...).then(resolve)) at the end of
   the current loop.
3. ontrack fires in the next loop.
```

Question: Is this on purpose? setRemoteDescription happens before tracks are fired.

If this isn't on purpose it's not obvious how to make the ontrack events fire first. If we don't queue them but fire them immediately, they're fired before we've completed all steps of setRemoteDescription, so we may be in an incomplete state. If we queue a task to resolve the promise then it would be possible to run a task in-between calling setRemoteDescription having had its effect and before the promise being resolved, which might yield unexpected behavior when examining the pc.

**Comment by taylor-b** [comment]
Good question. Firefox currently emits "track" before the promise resolves, and Chrome equivalently emits "addstream" before the promise resolves. That behavior makes sense to me; by the time a promise resolves, one would naturally assume everything associated with it is completed.

I think this was my fault: https://github.com/w3c/webrtc-pc/pull/1136

It was my shortsighted attempt to ensure that when each track event fires, the `stream` in the event has a full list of tracks. But I didn't think about how that would affect ordering relative to the promise resolving.

Assuming https://github.com/w3c/webrtc-pc/issues/1135 is still something we want to address, we could do it by saying "add the track event to a`[[pendingTrackEvents]]` internal slot" inside the "process the remote track" steps, and then after the "for each m= section" loop is finished, fire all pending track events. In other words, use an internal slot as a queue instead of the event queue.

**Comment by henbos** [comment]
Actually, I found that "addstream" is fired _after_ the promise's then-callback is invoked in Chrome for the exact same reasons as "track" in the spec and in my work-in-progress: it's queued to fire in the next loop.

A `[[pendingTrackEvents]]` would be a nice solution to this if we want a promise to be resolved after all events have guaranteed to be fired, since a fired track event does not tell you how many more events are pending.

**Comment by henbos** [comment]
As for the risk of inspecting the `pc` in an event between the effects of setRemoteDescription, but before the SRD's promise's then-callback, I think this is an inherent risk with promises that can't be avoided, regardless of how we write the spec, e.g:
```
Promise.resolve().then(function() {
  console.log("Remote streams: " + pc.getRemoteStreams().length);
});
pc.setRemoteDescription(sdpWithOneRemoteStream).then(function() {
  console.log("setRemoteDescription resolved!");
});
```
Because both promises are resolved before the event loop is empty, I would expect the callbacks to be invoked in order as soon as SRD is done and we get this output:
```
Remote streams: 1
setRemoteDescription resolved!
```

**Comment by taylor-b** [comment]
Yeah, I was wrong about Chrome's behavior; I got mixed up with something else. Hmm. That complicates things. But I still think it makes more sense for "track" events to fire before the SRD promise is resolved.

**Comment by henbos** [comment]
Related: With the recent https://github.com/w3c/webrtc-pc/pull/1402/files (@jan-ivar) we also add and remove tracks to the associated streams of the receiver. This results in addtrack and removetrack events firing, but note that this processing happens before we fire the `ontrack` and exposing the new streams, so there will be no chance to hook up anything to the streams' `onaddtrack`. Is this on purpose?

**Comment by jan-ivar** [comment]
(Cc @docfaraday whom I gave opposite advice before going on vk!)

[The promise guide](https://www.w3.org/2001/tag/doc/promises-guide#resolve-promise]), says *"If the algorithm using ["Resolve p with x" or "Reject p with r"] is running in parallel, the shorthands queue a task on p’s relevant settings object’s responsible event loop to call the stored function."*.

Which, since these steps of [SRD](http://w3c.github.io/webrtc-pc/#dom-rtcpeerconnection-setremotedescription) are in parallel, means that:
```
1. Process remote tracks.
   1a. Queue a task to fire an event named track.
2. Resolve the promise.
```
...will in fact resolve the promise *after* the event is fired. @domenic is that right?

Probably a good thing since SRD conditionally will also "await the completion of identity validation." even though this is written in prose and contradicts the prose directly above it where we return a promise solely from the [set-description](http://w3c.github.io/webrtc-pc/#set-description) sub-step.

**Comment by taylor-b** [comment]
@jan-ivar But the step of SRD that resolves the promise is not from a step run in parallel, it's from a queued task itself.

**Comment by jan-ivar** [comment]
@taylor-b Whoops, you're right. Please ignore my previous comment. I should have looked closer.

We probably also need to address then that the SRD promise may resolve much later if `setIdentityProvider` was called.

**Comment by jan-ivar** [comment]
> ensure that when each track event fires, the stream in the event has a full list of tracks

> This results in addtrack and removetrack events firing, but note that this processing happens before we fire the ontrack and exposing the new streams, so there will be no chance to hook up anything to the streams' onaddtrack. Is this on purpose?

If the streams (a track may be in more than one) are fully populated ahead of the `track` event, then it makes sense to me that users who hook the streams' `onaddtrack` or `onremovetrack` from within the first `ontrack` only hear of future renegotiations.

Note that [mediacapture](https://w3c.github.io/mediacapture-main/getusermedia.html#stream-inactive) appears to queue a task for each firing of `addtrack` or `removetrack` events on the stream, but the order should still hold, provided we continue to queue a task to fire the `track` event.
